build-tools/packages/build-infrastructure/src/utils.ts

```diff
@@ -0,0 +1,83 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+import execa from "execa";
+
+import { NotInGitRepository } from "./errors.js";
+
+/**
+ * Returns the absolute path to the nearest Git repository root found starting at `cwd`.
+ *
+ * @param cwd - The working directory to use to start searching for Git repositories. Defaults to `process.cwd()` if not
+ * provided.
+ *
+ * @throws A `NotInGitRepository` error if no git repo is found.
+ *
+ * @privateRemarks
+ * This function is helpful because it is synchronous. The SimpleGit wrapper is async-only.
+ */
+export function findGitRootSync(cwd = process.cwd()): string {
+	try {
+		const result = execa.sync("git", ["rev-parse", "--show-toplevel"], {
+			cwd,
+			encoding: "utf8",
+			// Ignore stdin but pipe (capture) stdout and stderr since git will write to both.
+			stdio: ["ignore", "pipe", "pipe"],
+		});
+
+		// If anything was written to stderr, then it's not a git repo.
+		if (result.stderr) {
+			throw new NotInGitRepository(cwd);
+		}
+
+		return result.stdout.trim();
+	} catch (error) {
+		throw new NotInGitRepository(cwd);
+	}
+}
+
+/**
+ * Returns the absolute path to the nearest Git repository found starting at `cwd`.
+ *
+ * @param cwd - The working directory to use to start searching for Git repositories. Defaults to `process.cwd()` if not
+ * provided.
+ *
+ * @privateRemarks
+ * This function is helpful because it is synchronous. The SimpleGit wrapper is async-only.
+ */
+export function isInGitRepositorySync(cwd = process.cwd()): boolean {
+	try {
+		const result = execa.sync("git", ["rev-parse", "--is-inside-work-tree"], {
+			cwd,
+			encoding: "utf8",
+			// Ignore stdin, pipe (capture) stdout, and ignore stderr
+			stdio: ["ignore", "pipe", "pipe"],
+		});
+
+		const isInWorktree = result.stdout.trim() === "true";
+		return isInWorktree;
+	} catch (error) {
+		return false;
+	}
```

Comment:
Same as previous comment, probably better to just let the error propagate?

Reply:
Hmmm, part of the intent of this function is that it doesn't throw - it just returns a boolean. Maybe a better name
would be `gitCommandsWorkInThisPath`? Ultimately this function is just trying to say "you can use Git here" or not.

Reply:
`gitCommandsWorkInThisPath` seems like a better name then. I'd still argue that any function that reaches out to external systems should be able to throw though. Say that the `--is-inside-work-tree` flag gets renamed, or rev-parse changes in some way that makes it throw _when invoked the way we do_; I think it's better behavior for this whole thing to blow (up to terminating the process) than for us to hide the error and have everything behave as if we're not in a git repo.

Also, just noticed the docs for this function are currently a copy of the docs for the one before, let's fix that too. If we don't let it throw, I'd also document that a return value of false could mean "not in a git repo" _or_ "cannot say for sure if we're in a git repo".

Reply:
I can coerce a NotInGitRepo error to a bool, but throw anything else.